### PR TITLE
scheduler: introduce the rank interface

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Ref #5691

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/schedule/schedulers/hot_region_rank_v1.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
+// Copyright 2017 TiKV Project Authors.
```
Comment (lhy1024) [inline pkg/schedule/schedulers/hot_region_rank_v1.go:1]:
```suggestion
// Copyright 2024 TiKV Project Authors.
```

**Review by lhy1024 — APPROVED** [review]

**Comment by okJiang** [comment]
What problem does this PR solve?🤔

File: pkg/schedule/schedulers/hot_region.go [hunk 2]
```diff
@@ -212,21 +212,17 @@ func (h *baseHotScheduler) summaryPendingInfluence(storeInfos map[uint64]*statis
...
 	for storeID, info := range storeInfos {
 		storeLabel := strconv.FormatUint(storeID, 10)
 		if infl := info.PendingSum; infl != nil && len(infl.Loads) != 0 {
 			utils.ForeachRegionStats(func(rwTy utils.RWType, dim int, kind utils.RegionStatKind) {
-				setHotPendingInfluenceMetrics(storeLabel, rwTy.String(), utils.DimToString(dim), infl.Loads[kind])
+				HotPendingSum.WithLabelValues(storeLabel, rwTy.String(), utils.DimToString(dim)).Set(infl.Loads[kind])
 			})
 		}
 	}
 }
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region.go:225]: It is better to wrap a function for these codes

File: pkg/schedule/schedulers/hot_region.go [hunk 3]
```diff
@@ -533,19 +535,8 @@ type balanceSolver struct {
 	maxPeerNum    int
 	minHotDegree  int
 
-	firstPriorityV2Ratios  *rankV2Ratios
-	secondPriorityV2Ratios *rankV2Ratios
-
 	// The rank correlation function used according to the version
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region.go:538]: Is this note still needed?

File: pkg/schedule/schedulers/hot_region.go [hunk 4]
```diff
@@ -557,9 +548,9 @@ func (bs *balanceSolver) init() {
 	bs.greatDecRatio, bs.minorDecRatio = bs.sche.conf.GetGreatDecRatio(), bs.sche.conf.GetMinorDecRatio()
 	switch bs.sche.conf.GetRankFormulaVersion() {
 	case "v1":
-		bs.initRankV1()
+		bs.rank = initRankV1(bs)
 	default:
-		bs.initRankV2()
+		bs.rank = initRankV2(bs, bs.greatDecRatio)
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region.go:553]: Why not input bs only?

File: pkg/schedule/schedulers/hot_region.go [hunk 5]
```diff
@@ -533,19 +535,8 @@ type balanceSolver struct {
...
-	isAvailable                 func(*solution) bool
-	filterUniformStore          func() (string, bool)
-	needSearchRevertRegions     func() bool
-	setSearchRevertRegions      func()
-	calcProgressiveRank         func()
-	betterThan                  func(*solution) bool
-	rankToDimString             func() string
-	checkByPriorityAndTolerance func(loads []float64, f func(int) bool) bool
-	checkHistoryLoadsByPriority func(loads [][]float64, f func(int) bool) bool
+	rank rank
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region.go:539]: How about this?

```suggestion
	rank
```

File: pkg/schedule/schedulers/hot_region_rank_v1.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+		return utils.DimToString(r.secondPriority)
+	case -1:
+		return utils.DimToString(r.firstPriority) + "-only"
+	default:
+		return "none"
+	}
+}
+
+func (*rankV1) needSearchRevertRegions() bool { return false }
+func (*rankV1) setSearchRevertRegions()       {}
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region_rank_v1.go:210]: 🤔
```suggestion
func (*rankV1) setSearchRevertRegions() {}
```

File: pkg/schedule/schedulers/hot_region_rank_v2.go [hunk 7]
```diff
@@ -59,7 +51,7 @@ type balanceChecker struct {
 // If the ratio is between the two, it is considered to be in the pre-balanced state.
 
 // TODO: Unified with stddevThreshold.
-type rankV2Ratios struct {
+type rankRatios struct {
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region_rank_v2.go:54]: please update all rankV2Ratios to rankRatios

File: pkg/schedule/schedulers/hot_region_rank_v2.go [hunk 8]
```diff
@@ -95,62 +87,78 @@ func newRankV2Ratios(balancedRatio, perceivedRatio, minHotRatio float64) *rankV2
...
+	default:
+		return r.checkByPriorityAndToleranceAnyOf(loads, f)
+	}
 }
 
 // pickCheckPolicyV2 will set checkByPriorityAndTolerance to the corresponding function.
 // Note: PolicyV2 will search more possible solutions than PolicyV1.
 // so it allows to schedule when any of the two dimensions is not balanced.
-func (bs *balanceSolver) pickCheckPolicyV2() {
+func (r *rankV2) checkHistoryLoadsByPriority(loads [][]float64, f func(int) bool) bool {
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region_rank_v2.go:134]: please update comment

File: pkg/schedule/schedulers/hot_region_rank_v2.go [hunk 9]
```diff
@@ -95,62 +87,78 @@ func newRankV2Ratios(balancedRatio, perceivedRatio, minHotRatio float64) *rankV2
...
-		bs.checkHistoryLoadsByPriority = bs.checkHistoryByPriorityAndToleranceAnyOf
+		return r.checkHistoryByPriorityAndToleranceAnyOf(loads, f)
 	}
 }
 
 // filterUniformStoreV2 filters stores by stddev.
 // stddev is the standard deviation of the store's load for all stores.
-func (bs *balanceSolver) filterUniformStoreV2() (string, bool) {
-	if !bs.enableExpectation() {
+func (r *rankV2) filterUniformStore() (string, bool) {
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region_rank_v2.go:145]: ditto

File: pkg/schedule/schedulers/hot_region_rank_v2.go [hunk 10]
```diff
@@ -191,48 +199,48 @@ func (bs *balanceSolver) setSearchRevertRegionsV2() {
 // |   isBetter                         | -4       | -3            | -2         |
 // |   isNotWorsened                    | -1       | 1             | 1          |
 // |   isWorsened                       | 0        | 1             | 1          |
-func (bs *balanceSolver) calcProgressiveRankV2() {
-	bs.cur.progressiveRank = 1
-	bs.cur.calcPeersRate(bs.firstPriority, bs.secondPriority)
-	if bs.cur.getPeersRateFromCache(bs.firstPriority) < bs.getMinRate(bs.firstPriority) &&
-		bs.cur.getPeersRateFromCache(bs.secondPriority) < bs.getMinRate(bs.secondPriority) {
+func (r *rankV2) calcProgressiveRank() {
```
Comment (okJiang) [inline pkg/schedule/schedulers/hot_region_rank_v2.go:202]: ditto

**Review by okJiang — APPROVED** [review]

File: pkg/schedule/schedulers/hot_region_rank_v2.go [hunk 11]
```diff
@@ -95,62 +87,76 @@ func newRankV2Ratios(balancedRatio, perceivedRatio, minHotRatio float64) *rankV2
...
 		futureChecker:  futureStateChecker,
 		currentChecker: currentStateChecker,
 		perceivedRatio: perceivedRatio, minHotRatio: minHotRatio}
 
 	return rs
 }
 
-func (bs *balanceSolver) initRankV2() {
-	bs.firstPriorityV2Ratios = newRankV2Ratios(bs.sche.conf.GetGreatDecRatio(), firstPriorityPerceivedRatio, firstPriorityMinHotRatio)
-	// The second priority is less demanding. Set the preBalancedRatio of the first priority to the balancedRatio of the second dimension.
```
Comment (HuSharp) [inline pkg/schedule/schedulers/hot_region_rank_v2.go:108]: Do we need to keep this comment?

**Review by HuSharp — APPROVED** [review]

**Pull request merged**